# Air Quality Dataset Aggregator

This notebook merges air quality data from multiple CSV files, filters by station ID, adds user-defined attributes, and splits the output into multiple files with a maximum of 10,000 rows each.

## Section 1: Import Required Libraries

In [1]:
import pandas as pd
import os
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')

## Section 2: Define User Input Parameters

Enter the station ID and the values for the new columns to be added to the dataset.

In [2]:
# Load the unified OSM features to get station data automatically
current_directory = os.getcwd()
unified_features_path = os.path.join(current_directory, "station_osm_features_unified.csv")

print("Loading unified OSM features...")
df_unified_features = pd.read_csv(unified_features_path)
print(f"✓ Loaded {len(df_unified_features)} stations from unified features")

# Identify all feature columns (exclude metadata columns)
metadata_cols = ['station_id', 'original_station_id', 'station_name', 'latitude', 'longitude', 
                 '_total_elements', '_unique_feature_types']
feature_columns = [col for col in df_unified_features.columns if col not in metadata_cols]

print(f"\nTotal feature columns: {len(feature_columns)}")
print(f"Feature columns: {feature_columns}")

print(f"\nFirst 5 stations (station info + first few features):")
display_cols = ['station_id', 'station_name'] + feature_columns[:5]
print(df_unified_features[display_cols].head())

print("\n✓ Ready to process all stations with ALL features automatically!")

Loading unified OSM features...
✓ Loaded 230 stations from unified features

Total feature columns: 26
Feature columns: ['aeroway', 'amenity', 'barrier', 'boundary', 'building', 'craft', 'emergency', 'healthcare', 'highway', 'historic', 'industrial', 'landuse', 'leisure', 'man_made', 'military', 'natural', 'office', 'place', 'power', 'public_transport', 'railway', 'shop', 'sport', 'tourism', 'water', 'waterway']

First 5 stations (station info + first few features):
  station_id                                     station_name aeroway  \
0      TN005                SIDCO Kurichi, Coimbatore - TNPCB     NaN   
1      KA012                    Urban, Chamarajanagar - KSPCB     NaN   
2      HR026                    MD University, Rohtak - HSPCB     NaN   
3      UP026  IESD Banaras Hindu University, Varanasi - UPPCB     NaN   
4      DL034                           Sirifort, Delhi - CPCB     NaN   

                                             amenity  \
0                                 

## Section 3: Load Station Hour Data

In [3]:
# Load station_hour.csv data
station_hour_path = os.path.join(current_directory, "station_hour.csv")

print(f"Loading station_hour.csv from: {station_hour_path}")
station_hour_df = pd.read_csv(station_hour_path)

print(f"✓ Station Hour DataFrame shape: {station_hour_df.shape}")
print(f"  Columns: {list(station_hour_df.columns)}")

Loading station_hour.csv from: /Users/likhithkanigolla/IIITH/code-files/Digital-Twin/Air-Quality-Dataset-Seggregator-DT/station_hour.csv
✓ Station Hour DataFrame shape: (2589083, 16)
  Columns: ['StationId', 'Datetime', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2', 'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket']


## Section 4: Process All Stations Automatically

This section automatically processes each station from the unified features dataset.

In [4]:
# Process all stations automatically
print("=" * 70)
print("AUTOMATED STATION PROCESSING")
print("=" * 70)

# Check if StationId column exists
if 'StationId' not in station_hour_df.columns:
    print("✗ StationId not found in station_hour.csv")
    print(f"Available columns: {list(station_hour_df.columns)}")
else:
    print("✓ StationId found in station_hour.csv")

# Get list of stations to process (only those in station_hour data)
available_stations = station_hour_df['StationId'].unique()
stations_to_process = df_unified_features[df_unified_features['station_id'].isin(available_stations)]

print(f"\nTotal stations in unified features: {len(df_unified_features)}")
print(f"Stations with air quality data: {len(stations_to_process)}")
print(f"Stations to process: {len(stations_to_process)}")

if len(stations_to_process) == 0:
    print("\n⚠️ Warning: No stations to process!")
else:
    print(f"\n✓ Ready to process {len(stations_to_process)} stations")

AUTOMATED STATION PROCESSING
✓ StationId found in station_hour.csv

Total stations in unified features: 230
Stations with air quality data: 96
Stations to process: 96

✓ Ready to process 96 stations


In [5]:
# Check unique StationId values in station_hour.csv
unique_station_ids = station_hour_df['StationId'].unique()

print("=" * 70)
print("UNIQUE STATION IDs IN STATION_HOUR.CSV")
print("=" * 70)
print(f"Total unique stations: {len(unique_station_ids)}")
print(f"\nStation IDs:")
print(sorted(unique_station_ids))

UNIQUE STATION IDs IN STATION_HOUR.CSV
Total unique stations: 110

Station IDs:
['AP001', 'AP005', 'AS001', 'BR005', 'BR006', 'BR007', 'BR008', 'BR009', 'BR010', 'CH001', 'DL001', 'DL002', 'DL003', 'DL004', 'DL005', 'DL006', 'DL007', 'DL008', 'DL009', 'DL010', 'DL011', 'DL012', 'DL013', 'DL014', 'DL015', 'DL016', 'DL017', 'DL018', 'DL019', 'DL020', 'DL021', 'DL022', 'DL023', 'DL024', 'DL025', 'DL026', 'DL027', 'DL028', 'DL029', 'DL030', 'DL031', 'DL032', 'DL033', 'DL034', 'DL035', 'DL036', 'DL037', 'DL038', 'GJ001', 'HR011', 'HR012', 'HR013', 'HR014', 'JH001', 'KA002', 'KA003', 'KA004', 'KA005', 'KA006', 'KA007', 'KA008', 'KA009', 'KA010', 'KA011', 'KL002', 'KL004', 'KL007', 'KL008', 'MH005', 'MH006', 'MH007', 'MH008', 'MH009', 'MH010', 'MH011', 'MH012', 'MH013', 'MH014', 'ML001', 'MP001', 'MZ001', 'OD001', 'OD002', 'PB001', 'RJ004', 'RJ005', 'RJ006', 'TG001', 'TG002', 'TG003', 'TG004', 'TG005', 'TG006', 'TN001', 'TN002', 'TN003', 'TN004', 'TN005', 'UP012', 'UP013', 'UP014', 'UP015', '

## Filter Unified Features for Stations with Air Quality Data

Create a filtered version of the unified features containing only the 110 stations that have air quality data.

In [6]:
# Filter unified features to only include stations with air quality data
print("Filtering unified features for stations with air quality data...")
print("=" * 70)

# Filter the unified features dataframe
df_unified_filtered = df_unified_features[df_unified_features['station_id'].isin(unique_station_ids)].copy()

print(f"Original unified features: {len(df_unified_features)} stations")
print(f"Filtered unified features: {len(df_unified_filtered)} stations")
print(f"Stations excluded: {len(df_unified_features) - len(df_unified_filtered)} stations")

# Save the filtered dataset
output_file = 'station_osm_features_filtered_110.csv'
df_unified_filtered.to_csv(output_file, index=False)

print(f"\n✓ Saved filtered dataset to '{output_file}'")
print(f"  - Rows: {len(df_unified_filtered)}")
print(f"  - Columns: {len(df_unified_filtered.columns)}")

# Show some statistics
print("\n" + "-" * 70)
print("FILTERED DATASET SUMMARY")
print("-" * 70)
print(f"\nFirst 5 stations in filtered dataset:")
display_cols = ['station_id', 'station_name', 'latitude', 'longitude']
print(df_unified_filtered[display_cols].head())

print(f"\nAll {len(df_unified_filtered)} filtered station IDs:")
print(sorted(df_unified_filtered['station_id'].tolist()))

Filtering unified features for stations with air quality data...
Original unified features: 230 stations
Filtered unified features: 96 stations
Stations excluded: 134 stations

✓ Saved filtered dataset to 'station_osm_features_filtered_110.csv'
  - Rows: 96
  - Columns: 33

----------------------------------------------------------------------
FILTERED DATASET SUMMARY
----------------------------------------------------------------------

First 5 stations in filtered dataset:
   station_id                         station_name   latitude  longitude
0       TN005    SIDCO Kurichi, Coimbatore - TNPCB  10.942451  76.978996
4       DL034               Sirifort, Delhi - CPCB  28.550425  77.215938
10      DL017             Lodhi Road, Delhi - IITM  28.588333  77.221667
11      MH014                 Worli, Mumbai - MPCB  18.993616  72.812811
12      TN004  Velachery Res. Area, Chennai - CPCB  13.005219  80.239812

All 96 filtered station IDs:
['AP001', 'AP005', 'AS001', 'BR005', 'BR006', 'BR00

## Section 5: Process Each Station with Features

This section loops through each station, filters the data, adds OSM features, and saves to chunks.

In [7]:
# Loop through each station and process
chunk_size = 10000
total_stations = len(stations_to_process)
processed_count = 0
failed_count = 0
total_files_created = 0

print(f"Starting automated processing of {total_stations} stations...")
print("=" * 70)

for idx, station_row in stations_to_process.iterrows():
    station_id = station_row['station_id']
    station_name = station_row['station_name']
    
    print(f"\n[{processed_count + 1}/{total_stations}] Processing: {station_id} - {station_name}")
    
    try:
        # Filter station_hour data for this station
        filtered_df = station_hour_df[station_hour_df['StationId'] == station_id].copy()
        
        if filtered_df.shape[0] == 0:
            print(f"  ⚠️ No data found for {station_id}, skipping...")
            failed_count += 1
            continue
        
        print(f"  Found {filtered_df.shape[0]} rows")
        
        # Add ALL feature columns from unified dataset
        for feature_col in feature_columns:
            # Get the feature value for this station (handle NaN values)
            feature_value = station_row[feature_col]
            if pd.isna(feature_value):
                feature_value = ""  # Empty string for missing features
            filtered_df[feature_col] = feature_value
        
        print(f"  Added {len(feature_columns)} feature columns")
        
        # Create output folder for this station
        output_folder = os.path.join(current_directory, "output", station_id)
        os.makedirs(output_folder, exist_ok=True)
        
        # Split into chunks and save
        total_rows = filtered_df.shape[0]
        num_chunks = (total_rows + chunk_size - 1) // chunk_size
        
        for i in range(num_chunks):
            start_idx = i * chunk_size
            end_idx = min((i + 1) * chunk_size, total_rows)
            chunk_df = filtered_df.iloc[start_idx:end_idx]
            
            file_number = i + 1
            filename = f"{station_id}_chunk_{file_number}.csv"
            filepath = os.path.join(output_folder, filename)
            
            chunk_df.to_csv(filepath, index=False)
            total_files_created += 1
        
        print(f"  ✓ Saved {num_chunks} chunk file(s) to output/{station_id}/")
        processed_count += 1
        
    except Exception as e:
        print(f"  ✗ Error processing {station_id}: {e}")
        failed_count += 1

# Final summary
print("\n" + "=" * 70)
print("PROCESSING COMPLETE!")
print("=" * 70)
print(f"Total stations processed: {processed_count}/{total_stations}")
print(f"Failed stations: {failed_count}")
print(f"Total files created: {total_files_created}")
print(f"\n✓ All output files saved in 'output/' folder")

Starting automated processing of 96 stations...

[1/96] Processing: TN005 - SIDCO Kurichi, Coimbatore - TNPCB
  Found 9229 rows
  Added 26 feature columns
  ✓ Saved 1 chunk file(s) to output/TN005/

[2/96] Processing: DL034 - Sirifort, Delhi - CPCB
  Found 32051 rows
  Added 26 feature columns
  ✓ Saved 4 chunk file(s) to output/DL034/

[3/96] Processing: DL017 - Lodhi Road, Delhi - IITM
  Found 24662 rows
  Added 26 feature columns
  ✓ Saved 3 chunk file(s) to output/DL017/

[4/96] Processing: MH014 - Worli, Mumbai - MPCB
  Found 9255 rows
  Added 26 feature columns
  ✓ Saved 1 chunk file(s) to output/MH014/

[5/96] Processing: TN004 - Velachery Res. Area, Chennai - CPCB
  Found 48192 rows
  Added 26 feature columns
  ✓ Saved 5 chunk file(s) to output/TN004/

[6/96] Processing: TN002 - Manali Village, Chennai - TNPCB
  Found 9072 rows
  Added 26 feature columns
  ✓ Saved 1 chunk file(s) to output/TN002/

[7/96] Processing: DL008 - DTU, Delhi - CPCB
  Found 48192 rows
  Added 26 featur

## Summary

This automated notebook:
1. ✓ Loads the unified OSM features dataset
2. ✓ Automatically identifies ALL feature columns (not just 6)
3. ✓ Loads station_hour.csv with air quality data
4. ✓ Processes each station automatically:
   - Filters air quality data by station ID
   - Adds ALL OSM feature columns from unified dataset
   - Splits data into chunks of 10,000 rows
   - Saves chunks to `output/[StationID]/` folder
5. ✓ Provides detailed progress and summary statistics

All output files are organized in the `output/` folder by station ID!